lmorpho/lsystem.cpp

```diff
@@ -0,0 +1,306 @@
+        branch_angle(P.branch_angle),
+        diam_child(P.diam_child_a, P.diam_child_r),
+        term_test(P.pterm_k1, P.pterm_k2),
+        branch_test(P.pbranch_ov_k1, P.pbranch_ov_k2,
+            P.pbranch_nov_k1, P.pbranch_nov_k2),
+        max_extent(P.max_extent)
+    {}
+
+    lsys_distribution diam_soma;
+    lsys_distribution n_tree;
+    lsys_distribution diam_initial;
+    lsys_distribution length_step;
+    lsys_distribution roll_initial;
+    lsys_distribution pitch_initial;
+    lsys_distribution roll_section;
+    lsys_distribution pitch_section;
+    lsys_distribution taper;
+    lsys_distribution roll_at_branch;
+    lsys_distribution branch_angle;
+    burke_correlated_r diam_child;
+    burke_exp_test term_test;
+    burke_exp2_test branch_test;
+    double max_extent;
+};
+
+struct section_tip {
+    section_point p = {0., 0., 0., 0.};
+    quaternion rotation = {1., 0., 0., 0.};
+    double somatic_distance = 0.;
+};
+
+struct grow_result {
+    std::vector<section_point> points;
+    std::vector<section_tip> children;
+    double length = 0.;
+};
+
+constexpr double deg_to_rad = 3.1415926535897932384626433832795l/180.0;
+
+template <typename Gen>
+grow_result grow(section_tip tip, const lsys_sampler& S, Gen &g) {
+    constexpr quaternion xaxis = {0, 1, 0, 0};
+    static std::uniform_real_distribution<double> U;
+
+    grow_result result;
+    std::vector<section_point>& points = result.points;
+
+    points.push_back(tip.p);
+    for (;;) {
+        quaternion step = xaxis^tip.rotation;
+        double dl = S.length_step(g);
+        tip.p.x += dl*step.x;
+        tip.p.y += dl*step.y;
+        tip.p.z += dl*step.z;
+        tip.p.r += dl*0.5*S.taper(g);
+        tip.somatic_distance += dl;
+        result.length += dl;
+
+        if (tip.p.r<0) tip.p.r = 0;
+
+        double phi = S.roll_section(g);
+        double theta = S.pitch_section(g);
+        tip.rotation *= rotation_x(deg_to_rad*phi);
+        tip.rotation *= rotation_y(deg_to_rad*theta);
+
+        points.push_back(tip.p);
+
+        if (tip.p.r==0 || tip.somatic_distance>=S.max_extent) {
+            return result;
+        }
+
+        if (S.branch_test(dl, tip.p.r, g)) {
+            double branch_phi = S.roll_at_branch(g);
+            double branch_angle = S.branch_angle(g);
+            double branch_theta1 = U(g)*branch_angle;
+            double branch_theta2 = branch_theta1-branch_angle;
+            auto r = S.diam_child(g);
+
+            tip.rotation *= rotation_x(deg_to_rad*branch_phi);
+
+            section_tip t1 = tip;
+            t1.p.r = t1.p.r * r.rho1;
+            t1.rotation *= rotation_y(deg_to_rad*branch_theta1);
+
+            section_tip t2 = tip;
+            t2.p.r = t2.p.r * r.rho2;
+            t2.rotation *= rotation_y(deg_to_rad*branch_theta2);
+
+            result.children = {t1, t2};
+            return result;
+        }
+
+        if (S.term_test(dl, tip.p.r, g)) {
+            return result;
+        }
+    }
```

Comment:
This is covered by the `lsys_distibution` parameters `max_extent` (stops them growing too long) and `max_sections` (stops excessive branching). A bad parameter set could still potentially lead to non-termination though.